File: apps/ios/Packages/Sources/ShadowChatListFeature/ChatListViewModel.swift
import Combine
import Foundation

@MainActor
public final class ChatListViewModel: ObservableObject {
    @Published public private(set) var state: ChatListState

    private let repository: ChatListRepository
    private let onRoomSelected: @MainActor (String) -> Void

    public init(
        repository: ChatListRepository,
        initialState: ChatListState = .loading,
        onRoomSelected: @escaping @MainActor (String) -> Void = { _ in }
    ) {
        self.repository = repository
        self.state = initialState
        self.onRoomSelected = onRoomSelected
    }

    public func send(_ event: ChatListEvent) {
        switch event {
        case .appeared:
            guard state == .loading else { return }
            Task { await load() }
        case .refreshRequested, .retryRequested:
            Task { await load() }
        case let .roomSelected(roomId):
            onRoomSelected(roomId)
        }
    }

    public func load() async {
        state = .loading

        do {
            let items = try await repository.loadChatList()
            state = items.isEmpty ? .empty : .loaded(items: items)
        } catch {
            state = .failed
        }
    }
}

File: apps/android/features/chatlist/src/main/java/com/shadowchat/features/chatlist/ChatListViewModel.kt
package com.shadowchat.features.chatlist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ChatListViewModel(
    private val repository: ChatListRepository,
) : ViewModel() {
    private val _state = MutableStateFlow<ChatListUiState>(ChatListUiState.Loading)
    val state: StateFlow<ChatListUiState> = _state.asStateFlow()

    fun onEvent(event: ChatListEvent) {
        when (event) {
            ChatListEvent.Appeared -> {
                if (_state.value == ChatListUiState.Loading) {
                    load()
                }
            }
            ChatListEvent.RefreshRequested,
            ChatListEvent.RetryRequested -> load()
            is ChatListEvent.RoomSelected -> Unit
        }
    }

    fun load() {
        _state.value = ChatListUiState.Loading

        viewModelScope.launch {
            _state.value = try {
                val items = repository.loadChatList()
                if (items.isEmpty()) {
                    ChatListUiState.Empty
                } else {
                    ChatListUiState.Loaded(items)
                }
            } catch (_: Throwable) {
                ChatListUiState.Error
            }
        }
    }
}
